File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/types/digitaltwin/ElementDescriptor.java
package org.opcfoundation.webserver.types.digitaltwin;

import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.UaInstanceNode;
import org.opcfoundation.webserver.digitaltwin.element.ElementType;

public class ElementDescriptor {
    private final String id;
    private final LocalizedText displayName;
    private final NodeId typeId;
    private final @Nullable UaInstanceNode instance;

    public ElementDescriptor(String id, UaInstanceNode submodel)
    {
        this.id = id;
        this.displayName = LocalizedText.NULL_VALUE;
        this.typeId = NodeId.NULL_VALUE;
        this.instance = submodel;
    }

    public ElementDescriptor(String id, LocalizedText displayName, ElementType elementType)
    {
        this.id = id;
        this.displayName = displayName;
        this.typeId = elementType.nodeId();
        this.instance = null;
    }

    public String getId() {
        return id;
    }

    public LocalizedText getDisplayName() {
        return displayName;
    }

    public NodeId getTypeId() {
        return typeId;
    }

    public @Nullable UaInstanceNode getInstance() {
        return instance;
    }
}

File: opcua-webservice-node/src/server/types/digitaltwin/ElementDescriptor.ts
import { UaLocalizedText, UaNodeId } from "opcua-webapi-ts";
import { UaInstanceNode } from "../../addressspace/nodes/UaInstanceNode";
import { UaObjectType } from "../../addressspace/nodes/UaObjectType";

export class ElementDescriptor {
    private readonly _id: string;
    private readonly _displayName: UaLocalizedText;
    private readonly _typeId: UaNodeId;
    private readonly _instance: UaInstanceNode | null;

    constructor(id: string, instance: UaInstanceNode);
    constructor(id: string, displayName: UaLocalizedText, elementType: UaObjectType);
    constructor(id: string, displayNameOrInstance: UaLocalizedText | UaInstanceNode, elementType?: UaObjectType) {
        let displayName = UaLocalizedText.nullText;
        let typeId = UaNodeId.nullNodeId;
        let instance: UaInstanceNode | null = null;

        if (displayNameOrInstance instanceof UaLocalizedText && elementType != null) {
            displayName = displayNameOrInstance;
            typeId = elementType.nodeId;
        }

        if (displayNameOrInstance instanceof UaInstanceNode) {
            this._instance = displayNameOrInstance;
        }

        this._id = id;
        this._displayName = displayName;
        this._typeId = typeId;
        this._instance = instance;
    }

    get id(): string {
        return this._id;
    }

    get displayName(): UaLocalizedText {
        return this._displayName;
    }

    get typeId(): UaNodeId {
        return this._typeId;
    }

    get instance(): UaInstanceNode | null {
        return this._instance;
    }
}